state-chain/pallets/cf-lending-pools/src/general_lending.rs

```diff
@@ -0,0 +1,1635 @@
+		{
+			if let Some(swap_progress) = T::SwapRequestHandler::abort_swap_request(*swap_request_id)
+			{
+				let excess_amount = match self.repay_principal(
+					*loan_id,
+					*to_asset,
+					swap_progress.accumulated_output_amount,
+					true, /* liquidation */
+				) {
+					Ok(LoanRepaymentOutcome::FullyRepaid { excess_amount }) => {
+						fully_repaid_loans.push(loan_id);
+						excess_amount
+					},
+					Ok(LoanRepaymentOutcome::PartiallyRepaid) => 0,
+					Err(_) => {
+						// On failure the full amount is unspent
+						swap_progress.accumulated_output_amount
+					},
+				};
+
+				if excess_amount > 0 {
+					T::Balance::credit_account(&self.borrower_id, *to_asset, excess_amount);
+				}
+
+				collateral_collected.push(AssetCollateralForLoan {
+					loan_id: *loan_id,
+					loan_asset: *to_asset,
+					collateral_asset: *from_asset,
+					collateral_amount: swap_progress.remaining_input_amount,
+				});
+			} else {
+				log_or_panic!("Failed to abort swap request: {swap_request_id}");
+			}
+		}
+
+		for loan_id in fully_repaid_loans {
+			self.settle_loan(*loan_id, true /* via liquidation */);
+		}
+
+		collateral_collected
+	}
+
+	/// Computes the total amount owed in account's loans in USD adjusting for the amount that will
+	/// be repaid by the collateral that has already been swapped into the loan asset.
+	pub fn total_owed_usd_value(&self) -> Result<AssetAmount, Error<T>> {
+		let total_owed = self
+			.loans
+			.values()
+			.map(|loan| loan.owed_principal_usd_value().ok())
+			.try_fold(0u128, |acc, x| acc.checked_add(x?))
+			.ok_or(Error::<T>::OraclePriceUnavailable)?;
+
+		let swapped_from_collateral = match &self.liquidation_status {
+			LiquidationStatus::NoLiquidation => 0,
+			LiquidationStatus::Liquidating { liquidation_swaps, .. } => {
+				let mut total_principal_usd_value_in_swaps = 0;
+				// If we are liquidating loans, we might already have some collateral swapped
+				// into the loan asset that will be used to repay the loan:
+				for (swap_request_id, LiquidationSwap { to_asset, .. }) in liquidation_swaps {
+					if let Some(swap_progress) =
+						T::SwapRequestHandler::inspect_swap_request(*swap_request_id)
+					{
+						total_principal_usd_value_in_swaps.saturating_accrue(usd_value_of::<T>(
+							*to_asset,
+							swap_progress.accumulated_output_amount,
+						)?);
+					} else {
+						log_or_panic!("Failed to inspect swap request: {swap_request_id}");
+					}
+				}
+				total_principal_usd_value_in_swaps
+			},
+		};
+
+		Ok(total_owed.saturating_sub(swapped_from_collateral))
+	}
+
+	/// Computes account's LTV (Loan-to-Value) ratio. Takes liquidations into account.
+	/// Returns error if oracle prices are not available, or if collateral is zero.
+	pub fn derive_ltv(&self) -> Result<FixedU64, Error<T>> {
+		let collateral = self.total_collateral_usd_value()?;
+		let principal = self.total_owed_usd_value()?;
+
+		if collateral == 0 {
+			return Err(Error::<T>::InsufficientCollateral);
+		}
```

Comment:
Can this ever happen? (would `log_or_panic` be more appropriate?)

And if it can happen, is an error the right return value? For example I can imagine we would want to top-up if the account has zero capital, but we would in fact return early (in `process_auto_topup`).

Reply:
This shouldn't be possible, but it is not locally obvious. 

> And if it can happen, is an error the right return value? 

I considered returning `FixedU64::max_value()`, but I think it is probably worse... If this returns error, then the topup won't work, but I think it is OK (if we got to a point where the user has loans but no collateral, something else must have gotten wrong, and we should investigate before taking more fund from the user). On the other hand, if the account has no loans and no collateral (but we failed to remove the account for some reason), it seems nicer to show LTV as `None` (what we would currently return in the RPC) instead of `max_value()`. 🤷 

I will add `log_or_panic`, but will leave the Error in for now.